torch/_inductor/fx_passes/misc_patterns.py

```diff
@@ -69,6 +69,52 @@ def randperm_index_replacement(x, slice_shape):
         scalar_workaround={"slice_shape": 42},
     )
 
+    # Float8 training patterns
+    E4M3_MAX_POS = torch.finfo(torch.float8_e4m3fn).max
+    E5M2_MAX_POS = torch.finfo(torch.float8_e5m2).max
+
+    def amax_with_scaling_pattern(tensor_x_inp, scale_x, IS_BWD):
+        tensor_x = tensor_x_inp.to(torch.float32) * scale_x
+        if IS_BWD:
+            tensor_x = tensor_x.clamp(min=-1 * E5M2_MAX_POS, max=E5M2_MAX_POS)
+            tensor_x = tensor_x.to(torch.float8_e5m2)
+        else:
+            tensor_x = tensor_x.clamp(min=-1 * E4M3_MAX_POS, max=E4M3_MAX_POS)
+            tensor_x = tensor_x.to(torch.float8_e4m3fn)
+        amax = torch.max(torch.abs(tensor_x_inp))
+        return (tensor_x, amax)
+
+    def amax_with_scaling_tiled_replacement(tensor_x_inp, scale_x, IS_BWD):
+        tensor_x = tensor_x_inp.to(torch.float32) * scale_x
+        if IS_BWD:
+            tensor_x = tensor_x.clamp(min=-1 * E5M2_MAX_POS, max=E5M2_MAX_POS)
+            tensor_x = tensor_x.to(torch.float8_e5m2)
+        else:
+            tensor_x = tensor_x.clamp(min=-1 * E4M3_MAX_POS, max=E4M3_MAX_POS)
+            tensor_x = tensor_x.to(torch.float8_e4m3fn)
+        amax_1 = torch.max(torch.abs(tensor_x_inp), dim=-1).values
+        amax = torch.max(amax_1)
+        return (tensor_x, amax)
+    
+    # The amax_with_scaling_pattern will also match dynamic scaling cases, we want to avoid that.
+    # `scale_x` of delayed scaling comes from the previous iteration (a `primals`), instead of from a function call.
+    def fp8_delayed_scaling_extra_check(match):
+        return match.kwargs["scale_x"].op != "call_function"
```

Comment:
not sure if this is going to survive future refactors in the torchao codebase, can we match on the pattern itself instead?

Reply:
Both dynamic and delayed scaling have the following pattern pieces:
```
        tensor_x = tensor_x_inp.to(torch.float32) * scale_x
        tensor_x = tensor_x.clamp(min=-1 * E4M3_MAX_POS, max=E4M3_MAX_POS)
        tensor_x = tensor_x.to(torch.float8_e4m3fn)
        amax = torch.max(torch.abs(tensor_x_inp))
```

The difference is the `scale_x` for delayed scaling comes from the sync_and_update_amax call (which is the input node to the forward/backward graph). But the `scale_x` of dynamic scaling is calculated from ` amax = torch.max(torch.abs(tensor_x_inp))` (a function call).

I am thinking to use another extra_check which checks if the input nodes of `scale_x` contains `tensor_x_inp`, which might work better to differentiate delayed and dynamic scaling. I'll update the pr if it works.

Reply:
hmm, I think if the logic here ends up being specific to how delayed scaling is implemented in torchao, then maybe that's evidence that torchao is a better place to put this code? Is there is an extension point in inductor where we can register this pass from torchao?

Reply:
Hi @eellison , do you have any suggestions here? maybe I can add a `custom_pattern_replacements` option in config? https://fburl.com/2f4y29hf

Reply:
@vkuzo I updated the logic of `extra_check`. It now checks whether `scale_x` is from `tensor_x_inp`. 
We might update how delayed scaling is implemented in torchao, but it should be always true that "In delayed scaling case, the input of `scale_x` does not contain `tensor_x_inp`", so the check less can be less coupled with torchao implementation.


I also got some performance number. 
As we discussed, we want to make "recompute fp8 weight and activation in backward" the default behavior, I also used these settings to test performance. I temporarily enforced them by using activation checkpointing (to recompute fp8 activation) and the config `force_recompute_fp8_weight_in_bwd`. The local script, D67517255.

```
Shape: 4096x4096x4096
Single linear layer
Before: 0.7000332674418591 ms
After: 0.6707110309278329 ms
Delta: 4.4%

With layer_norm:
Before: 0.753488252631584
After: 0.7374823118279574
Delta: 2.2%

Witi sigmoid:
Before: 0.7153208260869579
After: 0.6845765714285723
Delta: 4.5%
```